Makefile

```diff
@@ -1,27 +1,92 @@
-TOOLCHAIN_PATH=/opt/riscv64-unknown-elf-gcc-8.3.0-2020.04.1-x86_64-linux-ubuntu14/bin/
-FLASH_PATH=../flasher/ft2232h
-TOOLCHAIN_PREFIX=riscv64-unknown-elf
+#
+# Configure your RiscV toolchain path and prefix
+#
+TOOLCHAIN_PATH		?= /opt/riscv64-unknown-elf-gcc-8.3.0-2020.04.1-x86_64-linux-ubuntu14/bin/
+TOOLCHAIN_PREFIX	?= riscv64-unknown-elf-
 
-flash_tigard: ${PATTERN}.bin
-	PYTHONPATH=$(FLASH_PATH) python3 $(FLASH_PATH)/caravelflash/flash_util.py  --write $<
+# This path appears to be outside this project tree, so maybe instructions here to download this tool ?
+FLASH_PATH		?= ../flasher/ft2232h
 
-%.elf: %.c ../deps/sections.lds ../deps/crt0_vex.S
-	$(TOOLCHAIN_PATH)$(TOOLCHAIN_PREFIX)-gcc -I../deps -I../deps/generated/ -O0 -mabi=ilp32 -march=rv32i -D__vexriscv__ -Wl,-Bstatic,-T,../deps/sections.lds,--strip-debug -ffreestanding -nostdlib -o $@ ../deps/crt0_vex.S ../deps/isr.c ../deps/stub.c $<
-	${TOOLCHAIN_PATH}$(TOOLCHAIN_PREFIX)-objdump -D $@ > ${PATTERN}.lst
+
+ifeq ($(MAKELEVEL),0)
+ ifeq (deps,$(wildcard deps))
+# Top level only
+ROOTDIR			= $(CURDIR)
+SUBMODULES		= tt02-test
+ else
+# Subdir module is toplevel
+ROOTDIR			= $(PWD)/..
+ endif
+export ROOTDIR
+unexport SUBMODULES
+endif
+
+
+GENDIR  = -I$(ROOTDIR)/deps/generated
+INCDIRS = -I$(ROOTDIR)/deps $(GENDIR)
+
+GCC	= $(TOOLCHAIN_PATH)$(TOOLCHAIN_PREFIX)gcc
+OBJDUMP	= $(TOOLCHAIN_PATH)$(TOOLCHAIN_PREFIX)objdump
+OBJCOPY = $(TOOLCHAIN_PATH)$(TOOLCHAIN_PREFIX)objcopy
+
+# VexRiscV on Caravel
+CFLAGS_ARCH	=	-mabi=ilp32 -march=rv32i -D__vexriscv__ -mstrict-align
+
+#
+#CFLAGS_CODEGEN	= -O0
+# For caravel production reduce code size and amount of stack usage around function calls and
+#   set RiscV ABI minimum frame size (16 bytes) due to the limited RAM caravel has 1.5KiB.
+CFLAGS_CODEGEN	=	-Os -fomit-frame-pointer -mpreferred-stack-boundary=4
+
+CFLAGS		=	-Wall
+
+SRC_FILES	=	crt0_vex.S \
+			isr.c \
+			stub.c
+
+SRCS		=	$(addprefix $(ROOTDIR)/deps/,$(SRC_FILES))
+
+all: submodule-all
+
+
+submodule-all:
```

Comment:
could you comment on this?

Reply:
The first target in the Makefile is the `default target` which is usually the `all` target, this is a well known target name a convention with Makefile's.

It was not a good default to have `flash_foobar` the first and therefore default target, especially when the flashing process appears to need some other software that is outside this repo to work, which a 3rd party checking out this repo probably does not have installed yet.

On the basis the main goal of this repo is to build the firmware images.  The flashing is a secondary support goal.

But with wanting it to work within the sub-directory for (you) and from the top level to build everything (me and maybe a 3rd parties checkout expectation), the build needs to understand if it is building one thing (a subdir) or everything (toplevel).